lib/resty/healthcheck.lua

```diff
@@ -1458,53 +1415,36 @@ function _M.new(opts)
   if (self.checks.active.healthy.active or self.checks.active.unhealthy.active)
     and active_check_timer == nil then
 
-    self:log(DEBUG, "starting timer to check active checks")
-    local err
-    local check_healthcheck_timer, err = resty_timer({
+    self:log(DEBUG, "worker ", ngx_worker_id(), " (pid: ", ngx_worker_pid(), ") ",
+      "starting active check timer")
+    active_check_timer, err = resty_timer({
       recurring = true,
-      -- check if no worker is actively checking health status every 10 check
-      -- periods
-      interval = CHECK_INTERVAL * 10,
+      interval = CHECK_INTERVAL,
       jitter = CHECK_JITTER,
       detached = false,
       expire = function()
-        if self:get_periodic_lock() and not active_check_timer then
-          self:log(DEBUG, "worker ", ngx_worker_id(), " (pid: ", ngx_worker_pid(), ") ",
-                  "starting active check timer")
-          active_check_timer, err = resty_timer({
-            recurring = true,
-            interval = CHECK_INTERVAL,
-            detached = false,
-            expire = function()
-              self:renew_periodic_lock()
-              local cur_time = ngx_now()
-              for _, checker_obj in ipairs(hcs) do
-                if checker_obj.checks.active.healthy.active and
-                  (checker_obj.checks.active.healthy.last_run +
-                  checker_obj.checks.active.healthy.interval <= cur_time)
-                then
-                  checker_obj.checks.active.healthy.last_run = cur_time
-                  checker_callback(checker_obj, "healthy")
-                end
-
-                if checker_obj.checks.active.unhealthy.active and
-                  (checker_obj.checks.active.unhealthy.last_run +
-                  checker_obj.checks.active.unhealthy.interval <= cur_time)
-                then
-                  checker_obj.checks.active.unhealthy.last_run = cur_time
-                  checker_callback(checker_obj, "unhealthy")
-                end
-              end
-            end,
-          })
-          if not active_check_timer then
-            self:log(ERR, "Could not start active check timer: ", err)
+        local cur_time = ngx_now()
+        for _, checker_obj in ipairs(hcs) do
```

Comment:
The targets list is duplicated among all workers, so when a worker restart the list is not lost. The lock creates an active check "owner", and the `CHECK_INTERVAL * 10` is used to monitor if said owner is still running (it might have restarted/died). If we remove this lock, every worker will own the list and run the active health checks on every `CHECK_INTERVAL`. So, with this change every target would get _number of workers_ checks on every check interval configured for it, instead of only one.